File: pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

//declare functions which generate inputs for Coleman-Liau formula
int get_letters(string s);
int get_words(string s);
int get_sentences(string s);

int coleman_liau(int a, int b, int c);

int main(void)
{
    //generates the information req
    uired to calculate reading age
    string text = get_string("Text: ");
    int letters = get_letters(text);
    int words = get_words(text);
    int sentences = get_sentences(text);
    int grade = coleman_liau(letters, words, sentences);
    
    //test filters to bracket the reading ages as required
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 1 && grade < 16)
    {
        printf("Grade %i\n", grade);
    }
    else
    {
        printf("Grade 16+\n");
    }
}

//increments a counter for each valid alphabetical letter in a string
int get_letters(string s)
{
    int letters = 0;
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        if (s[i] >= 65 && s[i] <= 122)
        {
            letters++;
        }
        else
        {
            
        }
    }
    return letters;
}

//increments a counter for each non-space character following a space
int get_words(string s)
{
    int words = 0;
    
    //bool to tell if the program is "in the process" of counting one word
    bool count = false;
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        if (s[i] != ' ' && count == false)
        {
            words++;
            count = true;
        }
        //resets the count bool
        else if (s[i] == ' ')
        {
            count = false;
        }
        //catch-all for other non space characters
        else
        {
            
        }
    }
    return words;
}

//increments a counter each time a '.', '!' or '?' is detected
int get_sentences(string s)
{
    int sentences = 0;
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            sentences++;
        }
        else
        {
            
        }
    }
    return sentences;
}

//coleman liau index calculator uses floats to calculate accurately, then rounds to produce an int.
int coleman_liau(int a, int b, int c)
{
    float l = (float) a / (float) b * 100;
    float s = (float) c / (float) b * 100;
    int cl = roundf(0.0588 * l - 0.296 * s - 15.8);
    return cl;
}

File: pset6/readability.py
import cs50


# get letters iterates over the string returning a value for characters using isalpha()
def get_letters(text):
    letters = ""
    for i in range(len(text)):
        if text[i].isalpha():
            letters += text[i]
    return len(letters)
    

# if by iterating over a atring you find a non space character immediately following a space
# increment a counter
def get_words(text):
    words = 1  # assume one word as input
    for i in range(len(text)):
        if not text[i].isspace() and text[i-1].isspace():
            words += 1
    return words


# counts ., !, ?
def get_sentences(text):
    sentences = 0
    for i in range(len(text)):
        if text[i] in [".", "!", "?"]:
            sentences += 1
    return sentences


# coleman liau formula returning a rounded value
def coleman_liau(letters, words, sentences):
    L = letters / words * 100
    S = sentences / words * 100
    cl = 0.0588 * L - 0.296 * S - 15.8
    return round(cl)
       
        
text = cs50.get_string("Text: ")
cl = coleman_liau(get_letters(text), get_words(text), get_sentences(text))

if cl < 1:
    print("Before Grade 1")
elif cl >= 16:
    print("Grade 16+")
else:
    print(f"Grade {cl}")